### ISSUE TITLE
JIT: segfault in LSTM

### ISSUE THREAD

**Issue by tengerye** [pytorch/pytorch#46375]
## 🐛 Bug

Segfault error when tried to convert JIT on LSTM with `flatten_parameters()`.

## To Reproduce

```
class MyModel(nn.Module):
    def __init__(self):
        super(MyModel, self).__init__()
        self.rnn = nn.LSTM(10, 20, 2)

    def forward(self, x):
        self.rnn.flatten_parameters()
        y = self.rnn(x)
        return y

model = MyModel()
jit_model = torch.jit.script(model)
print(jit_model)
```

## Environment

PyTorch version: 1.5.0+cu101
Is debug build: False
CUDA used to build PyTorch: 10.1
ROCM used to build PyTorch: N/A

OS: Debian GNU/Linux 10 (buster) (x86_64)
GCC version: (Debian 8.3.0-6) 8.3.0
Clang version: Could not collect
CMake version: version 3.16.3

Python version: 3.7 (64-bit runtime)
Is CUDA available: False
CUDA runtime version: 10.1.243
GPU models and configuration: 
GPU 0: GeForce RTX 2080 Ti
GPU 1: GeForce RTX 2080 Ti
GPU 2: GeForce RTX 2080 Ti
GPU 3: GeForce RTX 2080 Ti

Nvidia driver version: 418.56
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.16.4
[pip3] torch==1.5.0+cu101
[pip3] torchtext==0.6.0
[pip3] torchvision==0.6.0+cu101
[conda] Could not collect

**Comment by gmagogsfm** [comment]
Hi @tengerye,

I can't repro your segfault issue with latest PyTorch. Instead I got a clear error message detailing why TorchScript failed. Could you migrate to latest PyTorch (1.7) and try again?

This is the error I got:
```
torch.jit.frontend.UnsupportedNodeError: GeneratorExp aren't supported:
  File "/home/ycao/miniconda3/envs/pytorch/lib/python3.8/site-packages/torch/nn/modules/rnn.py", line 138
        # alias would break the assumptions of the uniqueness check in
        # Module.named_parameters().
        unique_data_ptrs = set(p.data_ptr() for p in self._flat_weights)
                              ~ <--- HERE
        if len(unique_data_ptrs) != len(self._flat_weights):
            return
```

**Comment by tengerye** [comment]
@gmagogsfm Hi, thank you for your kind reply. But the error message still indicates the problem from a in-library module `rnn.py`.

**Comment by GeekOrangeLuYao** [comment]
> Hi @tengerye,
> 
> I can't repro your segfault issue with latest PyTorch. Instead I got a clear error message detailing why TorchScript failed. Could you migrate to latest PyTorch (1.7) and try again?
> 
> This is the error I got:
> 
> ```
> torch.jit.frontend.UnsupportedNodeError: GeneratorExp aren't supported:
>   File "/home/ycao/miniconda3/envs/pytorch/lib/python3.8/site-packages/torch/nn/modules/rnn.py", line 138
>         # alias would break the assumptions of the uniqueness check in
>         # Module.named_parameters().
>         unique_data_ptrs = set(p.data_ptr() for p in self._flat_weights)
>                               ~ <--- HERE
>         if len(unique_data_ptrs) != len(self._flat_weights):
>             return
> ```

Yes, I have the same question when I use `flatten_parameters`.  The error message is the same. I think the problem come from the flattern_parameters as the method it use `set` generator which is not supported in torchscript.

The code like this:
```python
import typing
import torch
import torch.nn as nn


class Model(nn.Module):
    def __init__(self,
                 input_dim: int,
                 output_dim: int,
                 constant_num: int) -> None:
        super().__init__()
        self.linear = nn.Linear(input_dim, output_dim)
        self.constant_list = [torch.rand(input_dim, output_dim) for _ in range(constant_num)]
    
    def forward(self,
                inputs: torch.Tensor,
                ) -> torch.Tensor:
        # s_list = [p for p in self.constant_list]
        # s = set(s_list)
        s = set(p for p in self.constant_list)
        return self.linear(inputs)

    
net = Model(30, 10, 5)
inputs = torch.ones((16, 123, 30))
scripted_model = torch.jit.script(net)
print(scripted_model.code)        
```

The code will get the error msg:
```
torch.jit.frontend.UnsupportedNodeError: GeneratorExp aren't supported:
  File "try_set_in_model.py", line 20
        # s_list = [p for p in self.constant_list]
        # s = set(s_list)
        s = set(p for p in self.constant_list)
               ~ <--- HERE
        return self.linear(inputs)
```

But when the code change to this:
```python
import typing
import torch
import torch.nn as nn


class Model(nn.Module):
    def __init__(self,
                 input_dim: int,
                 output_dim: int,
                 constant_num: int) -> None:
        super().__init__()
        self.linear = nn.Linear(input_dim, output_dim)
        self.constant_list = [torch.rand(input_dim, output_dim) for _ in range(constant_num)]
    
    def forward(self,
                inputs: torch.Tensor,
                ) -> torch.Tensor:
        s = set()
        for p in self.constant_list:
            s.add(p)
        return self.linear(inputs)

net = Model(30, 10, 5)
inputs = torch.ones((16, 123, 30))
scripted_model = torch.jit.script(net)
print(scripted_model.code)        
```

It can print the script code correctly:
```
def forward(self,
    inputs: Tensor) -> Tensor:
  s = ^set()()
  return (self.linear).forward(inputs, )
```

I think `pytorch` should check the method for using the `set` generator in the code. Anyway, I think you can change the **`rnn.py`** code from pytorch to solve this problem, if you can make sure the code will make **no mistakes** because change that code is **dangerous**.

**Comment by Jackiexiao** [comment]
as https://github.com/pytorch/audio/pull/1621#issue-938277015 say: 
> After checking with @gibiansky, flatten_parameters can be moved to init so now Encoder and Tacotron2 are torchscriptable.

**Comment by janeyx99** [comment]
TorchScript has been deprecated in favor of torch.export as of 2.11. https://pytorch.org/blog/pytorch-2-11-release-blog/ Closing this issue as we are not further developing on it--please ping if you believe this issue was erroneously closed.
